### PR TITLE
PRT-1040-add-policy-updater

### PR DESCRIPTION
**Pull request by ranlavanet**

### CODE REVIEW

File: protocol/rpcconsumer/rpcconsumer.go [hunk 1]
```diff
@@ -170,12 +171,20 @@ func (rpcc *RPCConsumer) Start(ctx context.Context, txFactory tx.Factory, client
 				return err
 			}
 			chainID := rpcEndpoint.ChainID
+			// create policyUpdaters per chain
+			if policyUpdater, ok := policyUpdaters[rpcEndpoint.ChainID]; ok {
+				policyUpdater.AddPolicySetter(chainParser, *rpcEndpoint)
```
Comment (omerlavanet) [inline protocol/rpcconsumer/rpcconsumer.go:176]: we are ignoring the error here, but this error will mean we aren't setting the policy updater, so let's either make it fatal, or overwrite the policy updater

File: protocol/statetracker/updaters/policy_updater.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	pu.policyUpdatables[endpoint.ApiInterface] = policyUpdatable
+	return nil
+}
+
+func (pu *PolicyUpdater) UpdaterKey() string {
+	return CallbackKeyForPolicyUpdate + pu.chainId
+}
+
+func (pu *PolicyUpdater) BuildPolicyMapAndSetPolicy(policyUpdatable PolicySetter, policy *plantypes.Policy, apiInterface string) error {
+	policyMap, err := policyUpdatable.BuildMapFromPolicyQuery(policy, pu.chainId, apiInterface)
```
Comment (omerlavanet) [inline protocol/statetracker/updaters/policy_updater.go:63]: why do we need two functions here, we can just do SetPolicy in the policyUpdatable, it's not like we are using the policy map

File: protocol/statetracker/updaters/policy_updater.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	utils.LavaFormatDebug("PolicyUpdater, fetching current policy and updating the effective policy", utils.LogAttr("epoch", epoch), utils.LogAttr("chainId", pu.chainId))
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+	policy, err := pu.policyFetcher.GetConsumerPolicy(ctx, pu.consumerAddress, pu.chainId)
+	if err != nil {
+		utils.LavaFormatError("could not get GetConsumerPolicy updated, did not update policy", err, utils.LogAttr("epoch", epoch))
+		return
+	}
+	for apiInterface, policyUpdatable := range pu.policyUpdatables {
+		err = pu.BuildPolicyMapAndSetPolicy(policyUpdatable, policy, apiInterface)
```
Comment (omerlavanet) [inline protocol/statetracker/updaters/policy_updater.go:87]: can just call it setPolicy

File: protocol/statetracker/updaters/event_tracker.go [hunk 4]
```diff
@@ -141,8 +143,13 @@ func (et *EventTracker) getLatestSpecModifyEvents(latestBlock int64) (updated bo
 	if et.latestUpdatedBlock != latestBlock {
 		return false, utils.LavaFormatWarning("event results are different than expected", nil, utils.Attribute{Key: "requested latestBlock", Value: latestBlock}, utils.Attribute{Key: "current latestBlock", Value: et.latestUpdatedBlock})
 	}
+	eventsListToListenTo := []string{
```
Comment (omerlavanet) [inline protocol/statetracker/updaters/event_tracker.go:146]: nice job this is an important fix~!

File: x/projects/types/types.go [hunk 5]
```diff
@@ -15,4 +15,5 @@ const (
 const (
 	AddProjectKeyEventName = "add_key_to_project_event"
 	DelProjectKeyEventName = "del_key_from_project_event"
+	PolicyUpdateEventName  = "policy_update_event"
```
Comment (omerlavanet) [inline x/projects/types/types.go:18]: unused

File: protocol/statetracker/updaters/state_query.go [hunk 6]
```diff
@@ -95,6 +95,7 @@ func (csq *StateQuery) GetDowntimeParams(ctx context.Context) (*downtimev1.Param
 
 type ConsumerStateQuery struct {
 	StateQuery
+	EpochStateQuery
```
Comment (omerlavanet) [inline protocol/statetracker/updaters/state_query.go:98]: is this used? we are using pairing updaters so i think this can be removed

**Review by omerlavanet — CHANGES_REQUESTED** [review]

**Comment by ranlavanet** [comment]
@omerlavanet I;ve merged fix-e2e-panic-issue here. so tests will pass. feel free to merge main here after to see the diff

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**